chainer/dataset/indexer.py

```diff
@@ -0,0 +1,191 @@
+import numpy
+import six
+
+
+class ExtractBySliceNotSupportedError(Exception):
+    pass
+
+
+class BaseIndexer(object):
+    """Base class for Indexer"""
+
+    def __getitem__(self, item):
+        raise NotImplementedError
+
+
+class BaseFeatureIndexer(BaseIndexer):
+
+    """Base class for FeatureIndexer
+
+    FeatureIndexer can be accessed by 2-dimensional indices, axis=0 is used for
+    dataset index and axis=1 is used for feature index.
+    For example, let `features` be the instance of `BaseFeatureIndexer`, then
+    `features[i, j]` returns `i`-th dataset of `j`-th feature.
```

Comment:
I do not fully understand the scope of this feature from this docstring. What is the expected behavior of an indexer attached to a dataset in each of the following cases?

Let `d` be a tuple dataset; then
- When `d` has multiple datasets one of which is again a tuple dataset (i.e., `d[i][j]` is a tuple)
- When `d` has a dataset which has variable shape (i.e., `d[0][0]` and `d[1][0]` has a different shape; consider the case of using variable length text input)